src/compiler/core.ts

```diff
@@ -899,20 +873,20 @@ export function compact<T>(array: readonly T[]): readonly T[]; // eslint-disable
 /** @internal */
 export function compact<T>(array: readonly T[]): readonly T[] {
     let result: T[] | undefined;
-    if (array) {
+    if (array !== undefined) {
         for (let i = 0; i < array.length; i++) {
             const v = array[i];
-            if (result || !v) {
-                if (!result) {
-                    result = array.slice(0, i);
-                }
+            // Either the result has been initialized (and is looking to collect truthy values separately),
+            // or we've hit our first falsy value and need to copy over the current stretch of truthy values.
+            if (result ?? !v) {
+                result ??= array.slice(0, i);
                 if (v) {
                     result.push(v);
                 }
             }

```

Comment:
This might be more readable as:
```suggestion
            if (v) {
              // If result has been initialized it is looking to collect truthy values separately.
              // Otherwise it will be initialized when we hit our first falsy value.
              result?.push(v);
            } else {
              // If this is our first falsy value, copy over the current stretch of truthy values.
              result ??= array.slice(0, i);
            }
```